<a href="https://colab.research.google.com/github/AylinNaebzadeh/Network-Visualization-Backend-Project/blob/main/99522185_Aylin_Naebzadeh_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
import pandas as pd
import networkx as nx
from pprint import pprint
import collections

In [2]:
data = pd.read_excel('edges.xlsx')


In [10]:
G = nx.from_pandas_edgelist(data, 'sourceNodeId', 'targetNodeId', create_using=nx.DiGraph())
print(f"THE NETWORK IS: {G}")

THE NETWORK IS: DiGraph with 2708 nodes and 10556 edges


In [11]:
# Read the data from the xlsx file into a DataFrame
labels_df = pd.read_excel('nodes.xlsx')

# Create a dictionary that maps node IDs to their labels
labels_dict = labels_df.set_index('NodeId')['Labels'].to_dict()

# Set the node attributes in the graph
nx.set_node_attributes(G, labels_dict, 'label')

In [12]:
print(f"THE NETWORK IS: {G}")

THE NETWORK IS: DiGraph with 2708 nodes and 10556 edges


In [21]:
pprint(f"THE NODES IN THE NETWORK ARE: {G.nodes()}")

('THE NODES IN THE NETWORK ARE: [35, 1033, 103482, 103515, 1050679, 1103960, '
 '1103985, 1109199, 1112911, 1113438, 1113831, 1114331, 1117476, 1119505, '
 '1119708, 1120431, 1123756, 1125386, 1127430, 1127913, 1128204, 1128227, '
 '1128314, 1128453, 1128945, 1128959, 1128985, 1129018, 1129027, 1129573, '
 '1129683, 1129778, 1130847, 1130856, 1131116, 1131360, 1131557, 1131752, '
 '1133196, 1133338, 1136814, 1137466, 1152421, 1152508, 1153065, 1153280, '
 '1153577, 1153853, 1153943, 1154176, 1154459, 116552, 12576, 128540, 132806, '
 '135130, 141342, 141347, 148170, 15670, 1688, 175291, 178727, 18582, 190697, '
 '190706, 1956, 197054, 198443, 198653, 206371, 210871, 229635, 231249, '
 '248425, 249421, 254923, 259701, 259702, 263279, 263498, 265203, 273152, '
 '27510, 28290, 286500, 287787, 28851, 289779, 289780, 289781, 307015, 335733, '
 '33904, 33907, 35061, 38205, 387795, 415693, 41714, 427606, 44368, 45599, '
 '46079, 46431, 486840, 48766, 503883, 503893, 513189, 54129, 54131, 5611

In [17]:
pprint(f"THE EDGES IN THE NETWORK ARE: {G.edges()}")

('THE EDGES IN THE NETWORK ARE: [(35, 1033), (35, 103482), (35, 103515), (35, '
 '1050679), (35, 1103960), (35, 1103985), (35, 1109199), (35, 1112911), (35, '
 '1113438), (35, 1113831), (35, 1114331), (35, 1117476), (35, 1119505), (35, '
 '1119708), (35, 1120431), (35, 1123756), (35, 1125386), (35, 1127430), (35, '
 '1127913), (35, 1128204), (35, 1128227), (35, 1128314), (35, 1128453), (35, '
 '1128945), (35, 1128959), (35, 1128985), (35, 1129018), (35, 1129027), (35, '
 '1129573), (35, 1129683), (35, 1129778), (35, 1130847), (35, 1130856), (35, '
 '1131116), (35, 1131360), (35, 1131557), (35, 1131752), (35, 1133196), (35, '
 '1133338), (35, 1136814), (35, 1137466), (35, 1152421), (35, 1152508), (35, '
 '1153065), (35, 1153280), (35, 1153577), (35, 1153853), (35, 1153943), (35, '
 '1154176), (35, 1154459), (35, 116552), (35, 12576), (35, 128540), (35, '
 '132806), (35, 135130), (35, 141342), (35, 141347), (35, 148170), (35, '
 '15670), (35, 1688), (35, 175291), (35, 178727), (35, 18582

In [18]:
pprint(f"THE ADJACENCY MATRIX OF THE NETWORK IS: {nx.adjacency_matrix(G).todense()}")

('THE ADJACENCY MATRIX OF THE NETWORK IS: [[0 1 1 ... 0 0 0]\n'
 ' [1 0 0 ... 0 0 0]\n'
 ' [1 0 0 ... 0 0 0]\n'
 ' ...\n'
 ' [0 0 0 ... 0 1 1]\n'
 ' [0 0 0 ... 1 0 0]\n'
 ' [0 0 0 ... 1 0 0]]')


In [19]:
pprint(f"THE NODES WITH THEIR LABELS ARE: {G.nodes.data()}")

("THE NODES WITH THEIR LABELS ARE: [(35, {'label': 'L2'}), (1033, {'label': "
 "'L2'}), (103482, {'label': 'L3'}), (103515, {'label': 'L2'}), (1050679, "
 "{'label': 'L2'}), (1103960, {'label': 'L2'}), (1103985, {'label': 'L2'}), "
 "(1109199, {'label': 'L2'}), (1112911, {'label': 'L2'}), (1113438, {'label': "
 "'L2'}), (1113831, {'label': 'L2'}), (1114331, {'label': 'L2'}), (1117476, "
 "{'label': 'L2'}), (1119505, {'label': 'L2'}), (1119708, {'label': 'L2'}), "
 "(1120431, {'label': 'L3'}), (1123756, {'label': 'L2'}), (1125386, {'label': "
 "'L7'}), (1127430, {'label': 'L2'}), (1127913, {'label': 'L2'}), (1128204, "
 "{'label': 'L2'}), (1128227, {'label': 'L2'}), (1128314, {'label': 'L2'}), "
 "(1128453, {'label': 'L2'}), (1128945, {'label': 'L2'}), (1128959, {'label': "
 "'L2'}), (1128985, {'label': 'L2'}), (1129018, {'label': 'L2'}), (1129027, "
 "{'label': 'L2'}), (1129573, {'label': 'L2'}), (1129683, {'label': 'L2'}), "
 "(1129778, {'label': 'L2'}), (1130847, {'label': 'L2'}), (1

Degree centrality is a measure of the importance of a node in a network. It is calculated as the number of edges connected to a node, normalized by the maximum possible degree in the graph. In other words, it measures how many neighbors a node has.

In your case, the top 5 nodes by degree centrality are nodes 35, 6213, 1365, 3229, and 910. These nodes have the highest number of edges connected to them, which means they have the most neighbors in the network.

One possible interpretation of this result is that these nodes may be important or influential in the network due to their high connectivity. They may act as hubs or bridges between different parts of the network.

The `nx.degree_centrality()` function in `networkx` calculates the degree centrality for each node in a graph. Degree centrality is a measure of the importance of a node in a network, calculated as the number of edges connected to a node, normalized by the maximum possible degree in the graph.

Here is how the `nx.degree_centrality()` function works behind the scenes:

1. The function takes as input a graph `G`.
2. It calculates the maximum possible degree in the graph, which is equal to `n-1` for an undirected graph and `n-1` for a directed graph, where `n` is the number of nodes in the graph.
3. It creates an empty dictionary to store the degree centrality values for each node.
4. It iterates over each node in the graph and calculates its degree centrality as the number of edges connected to the node divided by the maximum possible degree.
5. It stores the degree centrality value for each node in the dictionary.
6. It returns the dictionary containing the degree centrality values for all nodes in the graph.


In [23]:
# Calculate degree centrality
degree_centrality = nx.degree_centrality(G)

# Report the top 5 nodes based on each centrality metric
pprint('****************** TOP 5 NODES BY DEGREE CENTRALITY ******************')
for node in sorted(degree_centrality, key=degree_centrality.get, reverse=True)[:5]:
    pprint(f'{node}: {degree_centrality[node]}')

'****************** TOP 5 NODES BY DEGREE CENTRALITY ******************'
'35: 0.12412264499445881'
'6213: 0.05762837089028445'
'1365: 0.054673069818987806'
'3229: 0.04802364240857037'
'910: 0.03250831178426302'


Closeness centrality is a measure of the centrality of a node in a network. It is calculated as the reciprocal of the average shortest path length from a node to all other nodes in the graph. In other words, it measures how close a node is to all other nodes in the network.

In your case, the top 5 nodes by closeness centrality are nodes 35, 6213, 3229, 887, and 4584. These nodes have the highest closeness centrality values, which means they have the shortest average distance to all other nodes in the network.

One possible interpretation of this result is that these nodes may be important or influential in the network due to their central position. They may be able to quickly disseminate information or influence to other parts of the network due to their short average distance to all other nodes.

The `nx.closeness_centrality()` function in `networkx` calculates the closeness centrality for each node in a graph. Closeness centrality is a measure of the centrality of a node in a network, calculated as the reciprocal of the average shortest path length from a node to all other nodes in the graph.

Here is how the `nx.closeness_centrality()` function works behind the scenes:

1. The function takes as input a graph `G`.
2. It creates an empty dictionary to store the closeness centrality values for each node.
3. It iterates over each node in the graph and calculates its closeness centrality as follows:
    - It calculates the shortest path length from the node to all other nodes in the graph using the `nx.shortest_path_length()` function.
    - It calculates the average shortest path length by taking the mean of all shortest path lengths.
    - It calculates the closeness centrality as the reciprocal of the average shortest path length.
4. It stores the closeness centrality value for each node in the dictionary.
5. It returns the dictionary containing the closeness centrality values for all nodes in the graph.




In [25]:
# Calculate closeness centrality
closeness_centrality = nx.closeness_centrality(G)

pprint('****************** TOP 5 NODES BY CLOSENESS CENTRALITY ******************')
for node in sorted(closeness_centrality, key=closeness_centrality.get, reverse=True)[:5]:
    pprint(f'{node}: {closeness_centrality[node]}')

'****************** TOP 5 NODES BY CLOSENESS CENTRALITY ******************'
'35: 0.22276881556604902'
'6213: 0.2211907346794579'
'3229: 0.21982549145258112'
'887: 0.21601312745183981'
'4584: 0.21595173101580423'


Betweenness centrality is a measure of the centrality of a node in a network. It is calculated as the fraction of all shortest paths between pairs of nodes in the graph that pass through a given node. In other words, it measures how often a node acts as a bridge along the shortest path between two other nodes.

In your case, the top 5 nodes by betweenness centrality are nodes 35, 3229, 4330, 1365, and 6213. These nodes have the highest betweenness centrality values, which means they are on the highest fraction of shortest paths between pairs of nodes in the network.

One possible interpretation of this result is that these nodes may be important or influential in the network due to their position as bridges between different parts of the network. They may play a crucial role in connecting different communities or clusters within the network and facilitating the flow of information or influence between them. However, without more information about the context and meaning of the nodes and edges in your network, it is difficult to draw any specific conclusions.

The `nx.betweenness_centrality()` function in `networkx` calculates the betweenness centrality for each node in a graph. Betweenness centrality is a measure of the centrality of a node in a network, calculated as the fraction of all shortest paths between pairs of nodes in the graph that pass through a given node.

Here is how the `nx.betweenness_centrality()` function works behind the scenes:

1. The function takes as input a graph `G`.
2. It creates an empty dictionary to store the betweenness centrality values for each node.
3. It calculates the shortest paths between all pairs of nodes in the graph using the `nx.shortest_path()` function.
4. It iterates over each node in the graph and calculates its betweenness centrality as follows:
    - It initializes a counter for the node to zero.
    - It iterates over all pairs of nodes in the graph and checks if the node is on the shortest path between them.
    - If the node is on the shortest path between a pair of nodes, it increments the counter for that node by one.
    - It divides the counter for the node by the total number of pairs of nodes to get the fraction of shortest paths that pass through the node.
5. It stores the betweenness centrality value for each node in the dictionary.
6. It returns the dictionary containing the betweenness centrality values for all nodes in the graph.


In [27]:
# Calculate betweenness centrality
betweenness_centrality = nx.betweenness_centrality(G)

pprint('****************** TOP 5 NODES BY BETWEENNESS CENTRALITY ******************')
for node in sorted(betweenness_centrality, key=betweenness_centrality.get, reverse=True)[:5]:
    pprint(f'{node}: {betweenness_centrality[node]}')

'****************** TOP 5 NODES BY BETWEENNESS CENTRALITY ******************'
'35: 0.23248831450559276'
'3229: 0.12610085690638292'
'4330: 0.08934413819511756'
'1365: 0.08534091150396639'
'6213: 0.07637499735733927'


Eigenvector centrality is a measure of the centrality of a node in a network. It is calculated as the relative importance of a node based on its connections to other important nodes in the network. In other words, it measures how well-connected a node is to other well-connected nodes.

In your case, the top 5 nodes by eigenvector centrality are nodes 35, 82920, 85352, 210871, and 887. These nodes have the highest eigenvector centrality values, which means they are well-connected to other well-connected nodes in the network.

One possible interpretation of this result is that these nodes may be important or influential in the network due to their connections to other important nodes. They may be part of a core group of well-connected nodes that play a central role in the network.

The `nx.eigenvector_centrality()` function in `networkx` calculates the eigenvector centrality for each node in a graph. Eigenvector centrality is a measure of the centrality of a node in a network, calculated as the relative importance of a node based on its connections to other important nodes in the network.

Here is how the `nx.eigenvector_centrality()` function works behind the scenes:

1. The function takes as input a graph `G`.
2. It creates an adjacency matrix representation of the graph using the `nx.to_numpy_array()` function.
3. It calculates the eigenvector corresponding to the largest eigenvalue of the adjacency matrix using the `numpy.linalg.eig()` function.
4. It normalizes the eigenvector so that its elements sum to one.
5. It creates a dictionary that maps each node in the graph to its corresponding element in the normalized eigenvector.
6. It returns the dictionary containing the eigenvector centrality values for all nodes in the graph.

The eigenvector centrality value for each node represents its relative importance in the network based on its connections to other important nodes. Nodes with high eigenvector centrality values are well-connected to other well-connected nodes in the network.


In [28]:
# Calculate eigenvector centrality
eigenvector_centrality = nx.eigenvector_centrality(G)

pprint('****************** TOP 5 NODES BY EIGENVECTOR CENTRALITY ******************')
for node in sorted(eigenvector_centrality, key=eigenvector_centrality.get, reverse=True)[:5]:
    pprint(f'{node}: {eigenvector_centrality[node]}')

'****************** TOP 5 NODES BY EIGENVECTOR CENTRALITY ******************'
'35: 0.6542996956637285'
'82920: 0.11789362473669018'
'85352: 0.09924151614685083'
'210871: 0.0918477470705072'
'887: 0.09140594357270748'


In [32]:
# Calculate the degree distribution
degree_sequence = sorted([d for n, d in G.degree()], reverse=True)
degree_count = collections.Counter(degree_sequence)
deg, cnt = zip(*degree_count.items())

# Calculate the average clustering coefficient
avg_clustering_coefficient = nx.average_clustering(G)

# Report the results
pprint(f'Degree distribution: {degree_count}')
pprint(f'Average clustering coefficient: {avg_clustering_coefficient}')

('Degree distribution: Counter({4: 583, 6: 553, 2: 485, 8: 389, 10: 281, 12: '
 '131, 14: 82, 16: 57, 20: 26, 18: 25, 24: 18, 22: 14, 34: 8, 32: 7, 30: 6, '
 '28: 6, 38: 5, 26: 5, 46: 3, 42: 3, 36: 3, 64: 2, 60: 2, 336: 1, 156: 1, 148: '
 '1, 130: 1, 88: 1, 84: 1, 80: 1, 72: 1, 68: 1, 66: 1, 62: 1, 58: 1, 52: 1, '
 '44: 1})')
'Average clustering coefficient: 0.2406732985019372'
